packages/core/src/stateUtils.ts

```diff
@@ -1578,7 +1582,8 @@ export function resolveMicroTransition<
     currentState.context,
     new Set(prevConfig),
     machine,
-    _event
+    _event,
+    new Set(!currentState._initial ? [] : [machine.root])
```

Comment:
By moving `mutStatesToEnter` initialization to here we make it easier to add the root as a state that has to be entered during the fake~ initial transition. 

The usual expectation is that in a transition similar to this the `addAncestorStatesToEnter` should add all ancestors of the target state to `mutStatesToEnter` (if needed, ofc). However, this relies on `getProperAncestors` and that algorithm doesn't include the `toStateNode` in its result, so by using this logic alone we were not able to add the root state to the initial `mutStatesToEnter` and the root's invokes and entry actions rely on being included in this set for that initial transition.

SCXML doesn't account for this case in their algorithms because `<scxml/>` itself can't have `<invoke/>`, `<onentry/>`, etc. We are extending the spec by allowing those on our root state and that raises the need to make adjustments to the algorithms.

Reply:
This was going to be my main question - where this differs from the SCXML spec etc. Thanks for this!